source/credentials_provider_imds.c

```diff
@@ -565,86 +354,418 @@ static int s_make_imds_http_query(
         .request = request,
     };
 
-    struct aws_http_stream *stream =
-        impl->function_table->aws_http_connection_make_request(imds_user_data->connection, &request_options);
+    stream = impl->function_table->aws_http_connection_make_request(imds_user_data->connection, &request_options);
 
-    return stream == NULL ? AWS_OP_ERR : AWS_OP_SUCCESS;
+    if (!stream) {
+        goto on_error;
+    }
 
-on_error:
+    if (impl->function_table->aws_http_stream_activate(stream)) {
+        goto on_error;
+    }
+
+    return AWS_OP_SUCCESS;
 
+on_error:
+    impl->function_table->aws_http_stream_release(stream);
     aws_http_message_destroy(request);
 
     return AWS_OP_ERR;
 }
 
-static void s_imds_query_instance_role_credentials(struct aws_credentials_provider_imds_user_data *imds_user_data) {
-    AWS_FATAL_ASSERT(imds_user_data->connection);
+typedef void(imds_state_fn)(struct aws_credentials_provider_imds_user_data *);
+
+/* Make an http request to put a ttl and hopefully get a token back. */
+static void s_imds_query_token(struct aws_credentials_provider_imds_user_data *imds_user_data) {
+    AWS_PRECONDITION(imds_user_data->query_state == AWS_IMDS_QS_TOKEN_REQ);
+
+    struct aws_byte_cursor uri = aws_byte_cursor_from_string(s_imds_token_resource_path);
+
+    struct aws_http_header token_ttl_header = {
+        .name = aws_byte_cursor_from_string(s_imds_token_ttl_header),
+        .value = aws_byte_cursor_from_string(s_imds_token_ttl_default_value),
+    };
+
+    struct aws_http_header headers[] = {
+        token_ttl_header,
+    };
+
+    struct aws_byte_cursor verb = aws_byte_cursor_from_c_str("PUT");
+
+    imds_user_data->query_state = AWS_IMDS_QS_TOKEN_RESP;
+
+    if (s_make_imds_http_query(imds_user_data, &verb, &uri, headers, AWS_ARRAY_SIZE(headers))) {
+        imds_user_data->query_state = AWS_IMDS_QS_QUERY_NEVER_CLEARED_STACK;
```

Comment:
Could we set this before the call to s_make_imds_http_query? That way there is 0 chance of racing ever?

Reply:
Discussed this, it's not a problem.